### PR TITLE
zed: handle unions for map/array/sets

### PR DESCRIPTION
**Pull request by mattnibs**

### CODE REVIEW

File: runtime/expr/values.go [hunk 1]
```diff
@@ -275,48 +243,105 @@ type Entry struct {
...
+		// If there's only two types and one type is null, return single
+		// non-null type.
+		if types[0] == zed.TypeNull {
+			return types[1]
+		} else if types[1] == zed.TypeNull {
+			return types[0]
+		}
 	}
-	return ectx.CopyValue(zed.Value{m.typ, zed.NormalizeMap(bytes)})
+	return zctx.LookupTypeUnion(types)
```
Comment (nwt) [inline runtime/expr/values.go:346]: I think we should discard the null type whenever we have at least one other type.
```go
	 unique := types[:0]
	 for _, t := range zed.UniqueTypes(types) {
		if t != zed.TypeNull {
			unique = append(unique, t)
		}
	}
	if len(unique) == 0 {
		return zed.TypeNull
	}
	if len(unique) == 1 {
		return types[0]
	}
	return zctx.LookupTypeUnion(types)
```

File: runtime/expr/values.go [hunk 2]
```diff
@@ -178,95 +178,63 @@ func (r *recordSpreadExpr) invalidate(object map[string]column) {
 }
 
 type ArrayExpr struct {
-	zctx    *zed.Context
-	typ     *zed.TypeArray
-	builder *zcode.Builder
+	builder zcode.Builder
+	elems   collectionBuilder
 	exprs   []Evaluator
+	zctx    *zed.Context
```
Comment (nwt) [inline runtime/expr/values.go:184]: No need to adopt this if it doesn't warm your heart, but I find it helpful to group fields into stuff from the outside world that was passed to the constructor followed by a separate group of the purely internal stuff that's either created by the constructor or just left as the zero value.
```suggestion
	exprs []Evaluator
	zctx  *zed.Context

	builder zcode.Builder
	elems   collectionBuilder
```

**Review by nwt — APPROVED** [review]

File: runtime/expr/values.go [hunk 3]
```diff
@@ -178,95 +178,63 @@ func (r *recordSpreadExpr) invalidate(object map[string]column) {
...
-		bytes = []byte{}
+	if len(a.elems.types) == 0 {
+		return ectx.NewValue(a.zctx.LookupTypeArray(zed.TypeNull), []byte{})
+	}
+	it := a.elems.iter(a.zctx)
+	for !it.done() {
+		it.appendNext(&a.builder)
 	}
-	return ectx.NewValue(a.typ, bytes)
+	return ectx.NewValue(a.zctx.LookupTypeArray(it.typ), a.builder.Bytes()).Copy()
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply runtime/expr/values.go:207]: Why does this not need a copy? Won't the bytes get overwritten one the next value evaluated?

File: runtime/expr/values.go (hunk 1 above)
Comment (mattnibs) [reply runtime/expr/values.go:346]: What then would the selector of the null value be then?

File: runtime/expr/values.go [hunk 4]
```diff
@@ -275,48 +243,105 @@ type Entry struct {
...
+	c.types = append(c.types, val.Type)
+	c.bytes = append(c.bytes, val.Bytes)
+}
+
+func (c *collectionBuilder) iter(zctx *zed.Context) collectionIter {
+	// uniqueTypes must be copied since zed.UniqueTypes operates on the type
+	// array in place and thus we'll lose order.
+	for _, typ := range c.types {
+		c.uniqueTypes = append(c.uniqueTypes, typ)
+	}
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply runtime/expr/values.go:304]: doh

File: runtime/expr/values.go (hunk 1 above)
Comment (mattnibs) [reply runtime/expr/values.go:345]: Discussed this offline, will fix.

File: runtime/expr/values.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwt) [reply runtime/expr/values.go:208]: No, because at present, `zcode.Builder.Reset` sets `Builder.bytes` to nil.

**Pull request merged**